simulator/main.rs

```diff
@@ -152,74 +538,94 @@ fn main() {
+    // Shadow operation on the table
+    env.tables[table].rows.push(row.clone());
+
+    // Create a query that selects the row
+    let query = Query::Select {
+        table: env.tables[table].name.clone(),
+        predicate: Predicate::Eq(column.name.clone(), value),
     };
-    let rows = get_all_rows(env, conn, format!("SELECT * FROM {}", table_name).as_str())?;
 
-    let table = &env.tables[table];
-    compare_equal_rows(&table.rows, &rows);
-    Ok(())
+    // Get all rows
+    let rows = get_all_rows(env, conn, query.to_string().as_str()).unwrap();
+
+    // Check that the row is there
+    assert!(rows.iter().any(|r| r == &row));
 }
 
-fn do_write(env: &mut SimulatorEnv, conn: &mut Rc<Connection>) -> Result<()> {
-    let mut query = String::new();
+fn property_select_all(env: &mut SimulatorEnv, conn: &mut Rc<Connection>) {
+    // Get a random table
     let table = env.rng.gen_range(0..env.tables.len());
-    {
-        let table = &env.tables[table];
-        query.push_str(format!("INSERT INTO {} VALUES (", table.name).as_str());
-    }
 
-    let columns = env.tables[table].columns.clone();
-    let mut row = Vec::new();
+    // Create a query that selects all rows
+    let query = Query::Select {
+        table: env.tables[table].name.clone(),
+        predicate: Predicate::And(Vec::new()),
+    };
 
-    // gen insert query
-    for column in &columns {
-        let value = match column.column_type {
-            ColumnType::Integer => Value::Integer(env.rng.gen_range(i64::MIN..i64::MAX)),
-            ColumnType::Float => Value::Float(env.rng.gen_range(-1e10..1e10)),
-            ColumnType::Text => Value::Text(gen_random_text(env)),
-            ColumnType::Blob => Value::Blob(gen_random_text(env).as_bytes().to_vec()),
-        };
+    // Get all rows
+    let rows = get_all_rows(env, conn, query.to_string().as_str()).unwrap();
 
-        query.push_str(value.to_string().as_str());
-        query.push(',');
-        row.push(value);
+    // Check that all rows are there
+    assert_eq!(rows.len(), env.tables[table].rows.len());
+    for row in &env.tables[table].rows {
+        assert!(rows.iter().any(|r| r == row));
     }
+}
 
-    let table = &mut env.tables[table];
-    table.rows.push(row);
-
-    query.pop();
-    query.push_str(");");
+fn process_connection(env: &mut SimulatorEnv, conn: &mut Rc<Connection>) -> Result<()> {
+    if env.tables.is_empty() {
+        maybe_add_table(env, conn)?;
+    }
 
-    let _ = get_all_rows(env, conn, query.as_str())?;
+    match env.rng.gen_range(0..2) {
+        // Randomly insert a value and check that the select result contains it.
+        0 => property_insert_select(env, conn),
+        // Check that the current state of the in-memory table is the same as the one in the
+        // database.
+        1 => property_select_all(env, conn),
+        // Perform a random query, update the in-memory table with the result.
+        2 => {
```

Comment:
So now we dont have workload types. What I like of what I was doing is that there would be a Read/write/delete distribution where all the queries would derive from.

Reply:
I've added the workloads as one of the modes, let me know how that sounds, or if you think it's better to keep the workloads as the main testing system.

Reply:
Hmmm I don't think we want different modes. I think what we want is to randomize the environment first and then perform the simulation based on the parameters of the environment. For example at startup we decide workload distribution, maximum connections, number of tables we will want, how big we want texts to be, % of IO faults, etc... A different mode would mean that we fix the environment to something instead of having explicit logic if..else.